### ISSUE TITLE
Modifying openssl build parameters before building node?

### ISSUE THREAD

**Issue by RhianneMoll** [nodejs/help#2061]
Hi everyone,

I have to make SOAP calls over HTTPS to an ancient IIS6 production server (not under my control, unfortunately - it belongs to a government agency!) which only supports TLSv1 and three deprecated cipher suites: RC4-MD5, RC4-SHA and DES-CBC3-SHA. I had to rebuild OpenSSL with deprecated cipher support just to discover this.

My production environment is Node 10.16 on Ubuntu 18.04 and OpenSSL 1.1.1

In the wake of much dearly departed tooth enamel, I am now ready to build Node from the source in order to bake in support for these old cipher suites.

To build OpenSSL on its own with weak cipher support, one needs only to configure its makefile like so:

```./config enable-weak-ssl-ciphers -DOPENSSL_TLS_SECURITY_LEVEL=0```

What I want to know is, can I pass this to the node build configurator or must the included OpenSSL dependency in the deps folder be configured prior to building node? Specifically, must I configure *AND* build OpenSSL before building Node, or is it merely necessary to execute the aforementioned configure statement after which the OpenSSL build is performed as part of the Node build?

I'm also not sure whether the standard DEFAULT_CIPHERS tls constant will be altered as a result and I'd like to know if any changes to the source code are required as well.

Thanks in advance for any tips!
Rhianne

**Comment by bnoordhuis** [comment]
Node.js doesn't use openssl's config script (well... only during openssl upgrades; it's complicated) so you'll have to hack `deps/openssl/openssl.gyp` and `deps/openssl/openssl.gypi` to add the right defines and build the right files.

To be honest, I don't know what `enable-weak-ssl-ciphers` exactly enables and some quick spelunking through `Configure` didn't turn up anything so I'm afraid I can't give you specifics.

`tls.DEFAULT_CIPHERS` is just a fixed string. You can either update it in `lib/tls.js` or monkey-patch it at start-up.

> In the wake of much dearly departed tooth enamel

I feel your pain. ^^

**Comment by RhianneMoll** [comment]
:) Thanks for taking the time to respond. I discovered the constant in node_constants.h and modified it to suit, but even after a few spirit-crushing compile cycles I am not much nearer to success.

I've figured in the meantime that the build files for OpenSSL only apply when building it stand-alone and that I have to do something like what you've suggested. I'll shake the source tree some more and see what falls out.

This is turning into quite the nightmare... if not a character-building learning experience. :-p

**Comment by RhianneMoll** [comment]
SUCCESS!

It turned out (as it always does after the fact, dunnit?) to be relatively straight-forward. I replaced the included version of OpenSSL with the latest 1.1.1b source from openssl.org according to the instructions in the deps/openssl/config/README.md file.

I then altered /deps/openssl/config/Makefile and changed

```COPTS =  no-comp no-shared no-afalgeng enable-ssl-trace```

to

```COPTS =  no-comp no-shared no-afalgeng enable-ssl-trace enable-weak-ssl-ciphers -DOPENSSL_TLS_SECURITY_LEVEL=0```

... and then built the architecture includes by running make in the deps/openssl/config directory (I had to install nasm first and do a git init since I hadn't bothered to clone the repo).

That took care of the OpenSSL side of things. I then enabled the ciphers I needed by altering the string constant in node_constants.h and removing the bangs in front of MD5, RC4 and DES.

Then I built Node and it completed successfully. Many of the TLS tests in the testing suite failed however because they are clearly tailored for non-support of crappy cipher suites. Nevertheless, my test script to the IIS6 server connected successfully (with auto-negotiation too, I did not need to specify any cipher suites for the security context) and I was able to retrieve the WSDL.

Thanks for the help and the intuition pump!
Rhianne
